### PR TITLE
raft: remove flow control

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

### CODE REVIEW

File: raft/progress.go [hunk 1]
```diff
@@ -133,59 +49,19 @@ func (pr *Progress) optimisticUpdate(n uint64) { pr.Next = n + 1 }
...
-func (pr *Progress) needSnapshotAbort() bool {
-	return pr.State == ProgressStateSnapshot && pr.Match >= pr.PendingSnapshot
-}
-
 func (pr *Progress) String() string {
-	return fmt.Sprintf("next = %d, match = %d, state = %s, waiting = %v, pendingSnapshot = %d", pr.Next, pr.Match, pr.State, pr.IsPaused(), pr.PendingSnapshot)
+	return fmt.Sprintf("next = %d, match = %d", pr.Next, pr.Match)
 }
 
 type inflights struct {
```
Comment (Connor1996) [inline raft/progress.go]: we can remove this

File: raft/progress.go [hunk 2]
```diff
@@ -16,111 +16,27 @@ package raft
...
-
-	// Paused is used in ProgressStateProbe.
-	// When Paused is true, raft should pause sending replication message to this peer.
-	Paused bool
-	// PendingSnapshot is used in ProgressStateSnapshot.
-	// If there is a pending snapshot, the pendingSnapshot will be set to the
-	// index of the snapshot. If pendingSnapshot is set, the replication process of
-	// this Progress will be paused. raft will not resend snapshot until the pending one
-	// is reported to be failed.
-	PendingSnapshot uint64
```
Comment (Connor1996) [inline raft/progress.go]: I am worried about it breaks too many tests. Let's just remove `inflight` and `ProgressStateProbe` related code.

Comment (NingLin-P) [reply raft/progress.go]: Tests has been carefully handled (i.e remove or fix) and all tests remain pass.

File: raft/raft.go [hunk 3]
```diff
@@ -1032,39 +1002,15 @@ func stepLeader(r *Raft, m pb.Message) error {
...
 				if r.maybeCommit() {
 					if r.ShouldBcastCommit() {
 						r.bcastAppend()
 					}
-				} else if oldPaused {
-					// If we were paused before, this node may be missing the
-					// latest commit index, so send it.
-					r.sendAppend(m.From)
 				}
 				// We've updated flow control information above, which may
```
Comment (Connor1996) [inline raft/raft.go]: remove these comments

File: raft/raft.go [hunk 4]
```diff
@@ -499,9 +487,6 @@ func (r *Raft) sendAppend(to uint64) {
 // are undesirable when we're sending multiple messages in a batch).
 func (r *Raft) maybeSendAppend(to uint64, sendIfEmpty bool) bool {
 	pr := r.getProgress(to)
-	if pr.IsPaused() {
```
Comment (Connor1996) [inline raft/raft.go]: removing `SnapshotStatus` will make generate multiple snapshots every time when there is a snapshot is on the flight.

Comment (NingLin-P) [reply raft/raft.go]: Yes, you are right, `Snapshot` will be like `AppendLog` except contains a big `state machine` not `log entries`, should we keep flow control for snapshot?

File: raft/raft.go [hunk 5]
```diff
@@ -1112,28 +1053,7 @@ func stepLeader(r *Raft, m pb.Message) error {
...
-		// out the next MessageType_MsgAppend.
-		// If snapshot failure, wait for a heartbeat interval before next try
-		pr.pause()
 	case pb.MessageType_MsgUnreachable:
-		// During optimistic replication, if the remote becomes unreachable,
-		// there is huge probability that a MessageType_MsgAppend is lost.
-		if pr.State == ProgressStateReplicate {
-			pr.becomeProbe()
-		}
-		r.logger.Debugf("%x failed to send message to %x because it is unreachable [%s]", r.id, m.From, pr)
```
Comment (Connor1996) [inline raft/raft.go]: print this log

File: raft/raft.go (hunk 4 above)
Comment (Connor1996) [reply raft/raft.go]: alright, seems no correctness issue. @hicqu Is tikv snapshot mechanism based on the flow control of snapshot?

**Comment by NingLin-P** [comment]
> Please fix the conflict
Done.

**Pull request merged**